Update changefeed docs for OAuth parameters & Okta tutorial (#16761)
cockroachdb/docs: v23.1/connect-to-a-changefeed-kafka-sink-with-oauth-using-okta.md

---
title: Connect to a Changefeed Kafka sink with OAuth Using Okta
summary: Learn how to set up OAuth through Okta for a changefeed streaming to Kafka.
toc: true
docs_area: stream_data
---

> {% include_cached new-in.html version="v23.1" %} CockroachDB {{ site.data.products.enterprise }} changefeeds can stream change data out to [Apache Kafka](https://kafka.apache.org/) using OAuth authentication.

Reviewer: Can we link this back to more basic changefeed related detail?

Maybe just link the initial word changefeeds to 
 change-data-capture-overview.html , since that page, in turn, links to the Enterprise details?

Author: Yes!

{% include {{ page.version.version }}/cdc/oauth-description.md %}

In this tutorial, you will set up OAuth authentication for your Kafka changefeed sink using [Okta](https://www.okta.com/). You will create an OAuth 2.0 endpoint in Okta that will provide a temporary credential token that the changefeed will use to connect to your Kafka cluster.

An overview of the workflow involves:

1. Creating an Okta application integration and authorization server scope.
1. Configuring the Kafka cluster.
1. Streaming the changefeed to the Kafka cluster.

## Before you begin 

> You will need the following set up before starting this tutorial:

Reviewer: ```suggestion
Before starting this tutorial, you will need:
```

Author: Done!

> - A CockroachDB cluster. You can use a {{ site.data.products.db }} or {{ site.data.products.core }} cluster. If you are using {{ site.data.products.serverless }} or {{ site.data.products.dedicated }}, see the [Quickstart with CockroachDB](../cockroachcloud/quickstart.html) guide. For {{ site.data.products.core }} clusters, see the [install](install-cockroachdb-mac.html) page. 

Reviewer: ```suggestion
- A CockroachDB cluster. You can use a {{ site.data.products.db }} or {{ site.data.products.core }} cluster.
    - If you are using {{ site.data.products.db }} ({{ site.data.products.serverless }} or {{ site.data.products.dedicated }}), see the [Quickstart with CockroachDB](../cockroachcloud/quickstart.html) guide.
    - For {{ site.data.products.core }} clusters, see the [install](install-cockroachdb-mac.html) page. 
```

Forgetting our bullet indentation. Please confirm.

Author: Did it!

1. Log in to your Okta Developer account and navigate to **Applications** in the sidebar.

    In the dropdown, click **Applications** and then **Create App Integration**.

    On the **Create a new app integration** menu, select **API services** for OAuth 2.0 access tokens. Click **Next**.

    Add an **App integration name** and then **Save**.

    Once you have created the application integration, you'll create a scope for your authorization server.

1. Navigate to **Security** in the sidebar and then **API**. From here, you can add a new authorization server, or use the default authorization server. This tutorial will use the default server.

    Click on the name of your authorization server and then select the **Scopes** tab. 

    Select **Add Scope**:
    - Add `kafka` as the name for your scope and include an optional description. 
    - For **User Consent**, leave the default **Implicit** selected. 
    - Check the box to set this as the **Default Scope**. 
    - Optionally check **Include in public metadata** if you wish to include this scope in public metadata.

    Click **Create**.

> 

Reviewer: This might be easier to follow if you split it into two headings, enabling more numbered steps within each, e.g.

```
## Step 1. Create an Okta application
[...]
## Step 2. Configure a scope for your Okta application
```

Author: Adapted per this comment

Reviewer: Looks great

## Step 3. Update Kafka configuration

In this step, you will update configuration files in your Kafka cluster to set up the SASL mechanism for handling OAuth tokens. 

1. Move to the following directory:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    cd kafka_2.13-2.8.2/config
    ~~~

1. Open the `server.properties` file to edit:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    vi server.properties
    ~~~

    Add the following to the end of the `server.properties` file:

    {% include_cached copy-clipboard.html %}
    ~~~
    sasl.enabled.mechanisms=OAUTHBEARER
    sasl.mechanism.inter.broker.protocol=OAUTHBEARER
    security.inter.broker.protocol=SASL_PLAINTEXT
    listeners=PLAINTEXT://:9092,SASL_PLAINTEXT://:9093
    advertised.listeners=PLAINTEXT://:9092,SASL_PLAINTEXT://:9093
    auto.create.topics.enable=true

    listener.name.sasl_plaintext.oauthbearer.sasl.login.callback.handler.class=br.com.jairsjunior.security.oauthbearer.OauthAuthenticateLoginCallbackHandler
    listener.name.sasl_plaintext.oauthbearer.sasl.server.callback.handler.class=br.com.jairsjunior.security.oauthbearer.OauthAuthenticateValidatorCallbackHandler
    ~~~

> 

Author: Added

## Step 4. Prepare Kafka OAuth libraries

The SASL callback handlers added in [Step 3](#step-3-update-kafka-configuration) are part of a separate library that you will need to install.

1. Clone the library:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    git clone https://github.com/jairsjunior/kafka-oauth.git ~/kafka-oauth
    ~~~

1. Install [Apache Maven](https://maven.apache.org/what-is-maven.html), a tool for building Java-based projects:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    sudo apt install maven
    ~~~

1. Move to the `kafka-oauth` directory: 

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    cd ~/kafka-oauth
    ~~~

1. Run the following to compile and package the project:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    mvn package
    ~~~

## Step 5. Add OAuth environment variables

In this step, you'll export the environment variables that contain your OAuth client credentials. You will need the following information from your Okta developer account:

> - Your Okta developer account URL, for example `dev-12345678.okta.com`

Reviewer: Wondering if this should just be called "your Okta domain" https://developer.okta.com/docs/guides/find-your-domain/main/

Author: I think so, changing this

- The absolute paths to your: 
    - `kafka-oauth` directory
    - `kafka_2.13-2.8.2` directory (or your Kafka version)
- Your client ID and client secret from your Okta application (created in [Step 1](#step-1-create-an-okta-application-and-scope)). You will need to build the value for the variable as `{client ID}:{client secret}` and [base64 encode](https://www.base64encode.org/) the whole value.

> Export the following environment variables, or add them to your configuration file (e.g., `.bash_profile` or `.bashrc`) replacing the values between `{..}` with those mentioned in the preceding list:

Reviewer: ```suggestion
Export the following environment variables, or add them to your configuration file (e.g., `.bash_profile` or `.bashrc`), replacing the values in curly braces with those mentioned in the preceding list:
```

Author: Changed

export KAFKA_OPTS="-Djava.security.auth.login.config={absolute path to kafka_2.13-2.8.2}/config/kafka_server_jaas.conf"
~~~

## Step 6. Start your Kafka server

Ensure you're in the `kafka_2.13-2.8.2` directory, and then run the following:

1. Start the Zookeeper server in one terminal:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    ./bin/zookeeper-server-start.sh ./config/zookeeper.properties
    ~~~

1. Start the Kafka server in another terminal:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    ./bin/kafka-server-start.sh ./config/server.properties
    ~~~

1. In another separate terminal window, create a topic to consume:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    ./bin/kafka-topics.sh --bootstrap-server localhost:9092 --create --topic vehicles
    ~~~

    To consume messages for this topic, run:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    ./bin/kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic vehicles
    ~~~

## Step 7. Create a changefeed

In this step, you will create a changefeed authenticating with Okta.

The Kafka URI must follow this format:

~~~

> 'kafka://{kafka cluster address}:9093?topic_name={vehicles}&sasl_client_id={your client ID}&sasl_client_secret={your base64-encoded client secret}&sasl_enabled=true&sasl_mechanism=OAUTHBEARER&sasl_token_url={your Okta developer token URL}'

Author: Yes, very true. I have added this into the URI.

> 'kafka://{kafka cluster address}:9093?topic_name={vehicles}&sasl_client_id={your client ID}&sasl_client_secret={your base64-encoded client secret}&sasl_enabled=true&sasl_mechanism=OAUTHBEARER&sasl_token_url={your url-encoded Okta developer token URL}'

Reviewer: ```suggestion
kafka://{kafka cluster address}:9093?topic_name={vehicles}&sasl_client_id={your client ID}&sasl_client_secret={your base64-encoded client secret}&sasl_enabled=true&sasl_mechanism=OAUTHBEARER&sasl_token_url={your url-encoded Okta developer token URL}
```

Author: I'm actually going to add a copy-clipboard to this and keep the string quotes so that users can easily lift this and replace what they need to.

Reviewer: Makes sense - thanks.

~~~

Note the following:

- You can include your client ID directly (without any encoding).
- You must [base64 encode](https://www.base64encode.org/) your client secret.
- You must [URL encode](https://www.urlencoder.org/) your Okta developer token URL, e.g., `https://dev-12345678.okta.com/oauth2/default/v1/token` will encode to `https%3A%2F%2Fdev-12345678.okta.com%2Foauth2%2Fdefault%2Fv1%2Ftoken`.

Since this is a long URI to run, you can create an [external connection](create-external-connection.html) to represent this URI:

{% include_cached copy-clipboard.html %}
~~~ sql
CREATE EXTERNAL CONNECTION kafka_oauth AS 'kafka://{kafka cluster address}:9093?topic_name={vehicles}&sasl_client_id={your client ID}&sasl_client_secret={your base64-encoded client secret}&sasl_enabled=true&sasl_mechanism=OAUTHBEARER&sasl_token_url={your url-encoded Okta developer token URL}';
~~~

Create a changefeed that will emit messages to the topic consumer:

{% include_cached copy-clipboard.html %}
~~~ sql
CREATE CHANGEFEED FOR TABLE movr.vehicles INTO 'external://kafka_oauth' WITH updated, resolved;
~~~

> ~~~

Reviewer: The last code block and where its content appears isn't really explained. After you run `CREATE CHANGEFEED`, is there no substantial response in the SQL shell, and then here, you're representing the changefeed data that is being sent to Kafka? A one-liner of some kind between the code blocks should be sufficient.

Author: Add explanation here
